### PR TITLE
Add flymake kondor

### PR DESCRIPTION
**Pull request by theophilusx**
This PR adds flymake and flymake-kondor to add clj-kondo support to the clojure environment. It also adds 3 new key bindings -

SPC p d - display flymake diagnostics buffer
, g e - goto next flymake error/warning
, g E - goto previous flymake error/warning

You need to install clj-kondo for this mode to work. Will probably want to add some info to the manual, with a link to clj-kondo install instructions.

### CODE REVIEW

**Comment by plexus** [comment]
Thanks Tim, it'll be good to have some kind of answer for clj-kondo, but I'd prefer to not have it directly in corgi-clojure, so cave dwellers like myself can easily opt-out. Let's maybe move the flymake/kondor stuff into `corgi-kondo.el` (or `corgi-clojure-kondo.el`)?

- `SPC p d` what's the idea behind choosing this binding? `SPC p` is so for all for project/projectile related stuff. It seems these diagnostics are per-buffer? maybe something under `SPC b`, or something under `,` since it's fairly language-specific what that is going to look like
- `g e`/`g E` so far if we have a lowercase/uppercase version it's usually a variant command, e.g. `, s q` cider-quit vs `, s Q` corgi/cider-quit-all, or toggle soft vs hard word wrap. For previous/next symmetry I think we're better off leveraging the `]` and `[` keymaps, as discussed regarding evil-unimpaired, which actually has `] s` and `[ s` for prev/next flycheck error. My current thinking is to embrace `]` and `[` for these things in Corgi, but to do our own thing within that, maybe have `]/[ e` for prev/next error, that seems more intuitive.

It seems you're reverting some previous changes, e.g. regarding the bb-repl, I'm guessing that's not intentional?

**Comment by theophilusx** [comment]
I was considering corgi-clojure-extra to put this in, but if your happy with corgi-clojure-kondo.el, I'll put it in there. 

> SPC p d what's the idea behind choosing this binding?

Flymake can be (often is) configured to lint project wide, which is why I put the SPC p d binding. When you bring up the diagnostics buffer, you can move through the list of errors/warnings and hit enter to have the corresponding project file opened and cursor placed at that point.

> g e/g E

I used these bindings as they also mirror some other standard bindings in the motion state i.e. g t/g T for next/previous tab. (these days, I tend to use tabs more than frames i.e. :tabnew to create a new tab and g t/g T to move between tabs. 

I'm not 'wedded' to any specific binding scheme/pattern. My way of using Emacs is a little unusual - I'm vision impaired and an Emacspeak user. I use VERY large fonts, so rarely use split windows. As a consequence, I tend to modify key bindings to suit my special requirements. Unfortunately, this does make it hard to decide questions like these ones, so will defer to what you think would be most consistent. 

> It seems you're reverting some previous changes, e.g. regarding the bb-repl, I'm guessing that's not intentional?

Yep, complete screw up on my part. I forgot to merge the recent updates. 

I'll update and push the changes soon. Actually in isolation at present. Just got back from visiting my son and he tested positive on the day we left. Wife is now unwell and I'm trying to keep busy while waiting to see if the COVID axe falls!

**Comment by plexus** [comment]
> Actually in isolation at present.

Oh no! Keeping my fingers crossed for you!

File: corgi-clojure/corgi-clojure.el
Comment (plexus) [inline corgi-clojure/corgi-clojure.el]: Still some changes here that need to be reverted.

File: corgi-editor/corgi-editor.el
Comment (plexus) [inline corgi-editor/corgi-editor.el]: We don't want to force everyone to install this, especially if turning it on in every mode. I tried this out and started getting warnings in an emacs lisp file. I definitely do not want that.

Can we move this over to the kondo package, and only enable it in clojure/clojurescript buffers?

File: corgi-clojure-kondo/corgi-clojure-kondo.el
Comment (plexus) [inline corgi-clojure-kondo/corgi-clojure-kondo.el]: Having this comment here is confusing, since it's actually about `corgi-clojure`, not about `corgi-clojure-kondo`.

**Review by plexus — CHANGES_REQUESTED** [review]

File: corgi-editor/corgi-editor.el
Comment (plexus) [inline corgi-editor/corgi-editor.el]: What does this do?

**Comment by theophilusx** [comment]
My error. I forgot that as soon as I commit to the same branch it
automatically pushes through an updated PR. Not yet finished the
refactoring. Will let you know when I'm done (will see if I can
re-classify the PR as draft too). 

All the things you listed are on the todo list - just hadn't got there
yet.

**Comment by plexus** [comment]
Ok, sorry for the noise :)

File: corgi-clojure/corgi-clojure.el
Comment (theophilusx) [reply corgi-clojure/corgi-clojure.el]: Have restored corgi-clojure.el and corgi-editor.el from main and then added the flymake use-package entry to corgi-editor.el, so all should now be restored.

File: corgi-clojure-kondo/corgi-clojure-kondo.el
Comment (theophilusx) [reply corgi-clojure-kondo/corgi-clojure-kondo.el]: Removed

File: corgi-editor/corgi-editor.el
Comment (theophilusx) [reply corgi-editor/corgi-editor.el]: It is a backwards compatibility function used to provide diagnostics data in the same manner as the old version of flymake. I don't think we need it as we are not including any other legacy flymake configurations.

File: corgi-editor/corgi-editor.el
Comment (theophilusx) [reply corgi-editor/corgi-editor.el]: I will remove the add-hook to prog-mode-hook so that flymake is not loaded when you start a mode derived from prog-mode. However, I think we should leave the use-package stanza in corgi-editing as some people might want to use flymake with other modes (i.e. with javascirpt in particular). Without the add-hook prog-mode-hook expression, flymake should only be loaded either if corgi-clojoure-kondo is loaded (for clojure/clojurescript) or if flymake-mode is added to a mode hook (like js2-mode-hook) or if the user does M-x flymake-mode.

**Comment by theophilusx** [comment]
I think this PR is now ready for final review. 

- The missing commits have been re-introduced
- No longer adds flymake-mode to prog-mode-hook, so flymake not enabled in other modes (can be enabled by user if they want using Mx flymake-mode. Is enabled in clojure-mode if the load corgi-clojure-kondo. 
- Have not moved flymake to corgi-clojuire-kondo. Left it in corgi-edtining as it is a 'base' package. You typically add additional packages to specialise the mode for specific languages. Is not enabled unless user requests it. 
- Have not added key bindings for next/previous error. Looked at binding to [ e/] e, but those bindings not available due to evil-unimpaired. Wasn't sure where you want to bind them. They should be bound in at least clojure-mode. 

Note that I was unable to get things to work reliably using use-package. Don't understand why, but suspect it is something related to how we have corgi-packages structured. Was able to use use-package fine provided it was a module defined outside of corgi-packages. However, all works fine using straight-use-package. (I suspect some of the issue is related to the corgi-packages.el file and the fact I am using a branch other than 'main', though I did try modifying corgi-packages.el, but with no success). 

Once this pull request is merged, it will be necessary to update the manual and sample config file (will add use-package for corgi-clojure-kondo, but comment it out in sample config).

**Comment by plexus** [comment]
Looking into this some more, and only now realizing that there are two competing packages for integrating clj-kondo with emacs, [flymake-kondor](https://github.com/turbo-cafe/flymake-kondor) and [flycheck-clj-kondo](https://github.com/borkdude/flycheck-clj-kondo). The latter is what the clj-kondo docs recommend, and is authored by Michiel himself. It also has significantly more activity and popularity on Github. And I suppose there's a third option through clojure-lsp.

@theophilusx I guess you have some reasons to go with flymake? It seems flycheck generally has the momentum, and that it's considered more modern, faster, and more well maintained [[1]](https://www.reddit.com/r/emacs/comments/4ef3y2/flymake_vs_flycheck/) [[2]](https://www.masteringemacs.org/article/spotlight-flycheck-a-flymake-replacement).

One argument against flycheck is that its size is completely beyond normal proportions, clocking in at [12472 LoC](https://github.com/flycheck/flycheck/blob/master/flycheck.el#L12472). Flymake's proportions are modest in comparison, but still no liliputian at 1775 LoC.

Given this we definitely can not load either in `corgi-editor`. That's a core pacakge that pretty much every corgi user will be loading, we can't download and load packages there that for many people will be dead weight.

The biggest difficulty here is that I need to make an executive decision, because whatever we decide is going to have a long lasting impact on the project, and I don't feel comfortable making that decision right now. The only way that I can do that in an informed manner is to use flycheck/flymake with clj-kondo for a few months, and that's not something I'm able to invest that kind of time in right now.

What we can do is start by having a proper conversation about key bindings and signals. Conceptually what are the operations that are needed to support good use of linters? For instance: toggle linting on/off, go to next/previous error, open overview of linting errors... We can already include key bindings/signals for these even if we don't ship with support for them. What we're doing is creating an interface that third party packages can implement (by binding to signals).

Then the community can create flycheck-based or flymake-based linting packages (if there's enough interest for someone to do so), and people can seemlessly try out either.

@theophilusx how about I help you create a `theopilusx/corgi-flymake` package that sets up flymake+kondo?

Regarding `use-package`/`straight-use-package`, I believe you have to `(require 'use-package)` at the top, and also list it in your `package-requires`.

**Comment by theophilusx** [comment]
Arne Brasseur ***@***.***> writes:

> Looking into this some more, and only now realizing that there are two competing packages for integrating clj-kondo with emacs,
> flymake-kondor and flycheck-clj-kondo. The latter is what the clj-kondo docs recommend, and is authored by Michiel himself. It
> also has significantly more activity and popularity on Github. And I suppose there's a third option through clojure-lsp.
>
> @theophilusx I guess you have some reasons to go with flymake? It seems flycheck generally has the momentum, and that it's
> considered more modern, faster, and more well maintained [1] [2].
>

I'm not sure that information is up-to-date. There has been considerable
work put into updating flymake (starting wiht Emacs 27 and continued in
Emacs 28). The other reason I chose flymake was that it is a built-in
package which leverages off existing/available functionality. Flycheck
is larger and re-implements some of the functionality which already
exists (similar to what helm and ivy do wrt completion). 

I have used both flycheck and flymake and to be honest, have not
observed any clear advantage with either package - at least, not with
the languages I use (mainly Clojure and Javascript), with a bit of
latex, html and elisp. 

> One argument against flycheck is that its size is completely beyond normal proportions, clocking in at 12472 LoC. Flymake's
> proportions are modest in comparison, but still no liliputian at 1775 LoC.
>

Yes, flymake is smaller. However, I've not noticed any real performance
hit associated with elc and lines of code per se. I've certainly not
noticed any performance hit with either flycheck or flymake. 

As we are going to 'wrap' things inside some form of package i.e.
corgi-clojure-kondo, I don't think we need to worry too much. We can try
flymake and if it turns out to be unsatisfactory, we can just replace it
with flycheck. Both packages provide the same basic functionality and
they are really just wrappers around clj-kondo which does all the heavy
lifting. 

> Given this we definitely can not load either in corgi-editor. That's a core pacakge that pretty much every corgi user will be
> loading, we can't download and load packages there that for many people will be dead weight.
>
> The biggest difficulty here is that I need to make an executive decision, because whatever we decide is going to have a long
> lasting impact on the project, and I don't feel comfortable making that decision right now. The only way that I can do that in an
> informed manner is to use flycheck/flymake with clj-kondo for a few months, and that's not something I'm able to invest that kind
> of time in right now.
>
> What we can do is start by having a proper conversation about key bindings and signals. Conceptually what are the operations
> that are needed to support good use of linters? For instance: toggle linting on/off, go to next/previous error, open overview of
> linting errors... We can already include key bindings/signals for these even if we don't ship with support for them. What we're
> doing is creating an interface that third party packages can implement (by binding to signals).

I think those are the basics and all that is needed. 

If you can identify the appropriate interface 'slots' that would greatly
help. As mentioned previously, the [ and ] bindings would be good, but
most ot the available (shallow) bindings are already used by
evil-unimpaired. I guess a decision is needed on whether to continue to
use unimpaired, in which case we will need to identify other bindings,
or we remove unimpaired and use [ e and ] e. 

> @theophilusx how about I help you create a theopilusx/corgi-flymake package that sets up flymake+kondo?

Yes, that is fine. I actually had a separate package to begin with and
only did the more tightly 'bound' version after you indicated that you
thought adding it as part of corgi proper was warranted. I'm already
creating other packages which I planned to host on my github account
anyway. Have already started working on a corgi-javascript, corgi-eglot
(LSP support), corgi-spelling (ispeall/flyspell), and corgi-org (this
one is mainly about key bindings). 

>
> Regarding use-package/straight-use-package, I believe you have to (require 'use-package) at the top, and also list it in
> your package-requires.

Hmm, OK, I'll give that a go. Had assumed that as they are already
required by packages loaded before this one, it wouldn't make a
difference. Note that when I had it as a separate package (outside of
corgi-packages), use-package worked fine - it was only when I moved it
to be part of corgi-packages that I couldn't get use-pacakge to work
reliably. 

The biggest challenge I find at this time is how to install package
specific signal definitions (or key bindings for that matter). I'm
wondering if we actually need to have a couple of custom variables i.e.
corgi-key-files and corgi-signal-files? The issue I run into is that
corkey/load-and-watch doesn't add to definitions, it reloads the
definitions. 

Consider where I'm loading 3 additonal packages p1, p2 and p3. Each has
signals and possibly additional key bindings I which should be added if
each of these packages is loaded. How can I tell corkey about these
additional key and signal files in an incremental and non-distructive
manner? Currently, the only way I can see to do it is for the user to
put a call to corkey/load-and-watchm, including 2 lists for the set of
key and signal files (including the defaults and those for the
additional package they ahve loaded). 

Even if none of the packages create any new bindings, they will each
want to set signals and I'm having trouble working out a clean way to do
this which minimises the amount of internal knowledge the user needs to
have about corkey and the additional packages they are loading. I'd
prefer to be able to have use-package blocks that setup the signals (and
possibly additonal bindings - especially mode specific once bound to the
',' leader) so that all the user has to do is load the 'outer' package. 

One of the advantages of corkey is in having all your key bindings and
signals defined in just two files (ignoring custom user ones). Maybe it
would also be good to have an interface with would display a virtual
keys and virtual signals file which is the accumulation of all key and
signal definitions, regardless of source?

I'll close the PR and create a corgi-clojure-kondo package and make it
available on my github account. In the meantime, if you can help with
the key bindings stuff, that would be great. 

By the way, Have you ever looked at this package -

https://github.com/jpe90/emacs-clj-deps-new

Apparently, it provides a 'magit like' interface for creating new
clojure projects. I've only just heard about it. I much prefer CLI tools
over lein these days. Might be interesting. Since the addition of
trransient-mode, I've thought about whether it would be useful to have
some interactive interface over your syste+project deps.edn file (not
sure about you, but my deps.edn has gotten quite large and it has some
really useful things. However, many of them I use so rarely, I nearly
always have to go back to look at the definition and comments to jog my
memory on how to use it)

**Comment by theophilusx** [comment]
Closing this PR. Will implement as a separate package outside of corgi.

**Pull request closed without merging**
